### PR TITLE
feat: metrics logger

### PR DESCRIPTION
**Pull request by huwenchao**
**What type of PR is this?**

A logger and metrics implementation.

**What this PR does / why we need it**:

Implement a crate to make log and add metrics easier.

**Special notes for your reviewer**:

Check `README.md` for more information.

### CODE REVIEW

**Comment by huwenchao** [comment]
usage example: https://github.com/huwenchao/muta/commit/7c04ff3e906ab5e155c8258b63195740f9d1f0f5

**Comment by yejiayu** [comment]
So, what is the format of the output?

File: common/logger/log.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# This file is yaml style config, can make testing the logger more easily.
```
Comment (yejiayu) [inline common/logger/log.yml]: Where is it used?

File: common/logger/src/lib.rs [hunk 2]
```diff
@@ -1,73 +1,90 @@
...
-        trace!("Hello {}", "World!");
-        trace!("Hell{} {}", "o", "World!");
-        trace!("Hell{} World", 0);
-    }
+    let metrics_logger = Logger::builder().additive(false).appender("metrics").build(
+        "metrics",
+        if metrics {
+            LevelFilter::Trace
+        } else {
+            LevelFilter::Off
```
Comment (yejiayu) [inline common/logger/src/lib.rs]: Does this mean that `trace` information is lost?

The PR may be missing some necessary instructions.

Comment (huwenchao) [reply common/logger/src/lib.rs]: > Does this mean that trace logs will be discarded?

No, only the `metrics` target is discarded.

> The PR may be missing some necessary instructions.

That makes sense. I'll add a readme file.

File: common/logger/log.yml (hunk 1 above)
Comment (huwenchao) [reply common/logger/log.yml]: It is the yaml style of the rust code in `init` function in `lib.rs`, not used here, but makes it easier to understand the code and test new logger config.

If someone wants to have more customized logger config other than that in the `config.toml` file, he/she may need this.

File: common/logger/Cargo.toml [hunk 3]
```diff
@@ -7,5 +7,6 @@ edition = "2018"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-env_logger = "0.6"
 log = "0.4"
+log4rs = "0.8"
+serde_json = "1.0"
```
Comment (yejiayu) [inline common/logger/Cargo.toml]: newline

File: common/logger/log.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  appenders:
+  - console
+  - file
+
+loggers:
+  metrics:
+    level: trace
+    appenders:
+    - metrics
+    additive: false
```
Comment (yejiayu) [inline common/logger/log.yml]: newline

File: devtools/chain/config.toml [hunk 5]
```diff
@@ -41,3 +41,11 @@ precommit_denominator = 30
...
 [executor]
 light = false
+
+[logger]
+filter = "info"
+log_to_console = true
+console_show_file_and_line = false
+log_path = "logs/"
+log_to_file = true
+metrics = true
```
Comment (yejiayu) [inline devtools/chain/config.toml]: newline

File: common/logger/Cargo.toml (hunk 3 above)
Comment (huwenchao) [reply common/logger/Cargo.toml]: > newline

I wondered why newline is necessary, did some search and found this <https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline>.
 Pretty interesting.

**Pull request merged**